sdk/storage/Azure.Storage.Common/tests/LazyLoadingReadOnlyStreamTests.cs

```diff
@@ -0,0 +1,102 @@
+                        .Returns(new DownloadedContent(new MemoryStream(data, (int)range.Offset, (int)(range.Length ?? (data.Length - range.Offset)))));
+                    mockResponse.Setup(r => r.GetRawResponse())
+                        .Returns(() => new MockResponse(201).AddHeader("Content-Range", $"bytes {range.Offset}-{range.Offset + range.Length - 1}/{data.Length}"));
+                    return Task.FromResult(mockResponse.Object);
+                });
+
+            return downloadMock;
+        }
+
+        // int just fills the generic requirements, we don't care about data type and go straight for the raw response
+        private static Mock<LazyLoadingReadOnlyStream<int>.GetPropertiesAsync> GetPropertiesBehavior(int totalResourceLength)
+        {
+            var propertiesMock = new Mock<LazyLoadingReadOnlyStream<int>.GetPropertiesAsync>();
+            propertiesMock.Setup(_ => _(It.IsAny<bool>(), It.IsAny<CancellationToken>()))
+                .Returns<bool, CancellationToken>((value, cancellationToken) =>
+                {
+                    var mockResponse = new Mock<Response<int>>(MockBehavior.Loose);
+                    mockResponse.Setup(r => r.GetRawResponse())
+                        .Returns(() => new MockResponse(200).AddHeader("Content-Length", totalResourceLength.ToString()));
+                    return Task.FromResult(mockResponse.Object);
+                });
+
+            return propertiesMock;
+        }
+
+        [TestCase(0, Constants.KB, Constants.KB)]
+        [TestCase(0 * Constants.KB, Constants.KB, 3 * Constants.KB)]
+        [TestCase(1 * Constants.KB, Constants.KB, 3 * Constants.KB)]
+        [TestCase(2 * Constants.KB, Constants.KB, 3 * Constants.KB)]
+        public async Task AvoidRebuffer(int offset, int bufferSize, int dataSize)
+        {
+            if (offset < 0 || offset + bufferSize > dataSize)
+            {
+                throw new ArgumentException("Bad test definition.");
+            }
+
+            // Arrange
+            byte[] data = TestHelper.GetRandomBuffer(dataSize);
+
+            var downloadMock = GetDownloadBehavior(data);
+            var propertiesMock = GetPropertiesBehavior(data.Length);
+            var readStream = new LazyLoadingReadOnlyStream<int>(
+                downloadMock.Object,
+                propertiesMock.Object,
+                hashingOptions: default,
+                allowModifications: false,
+                initialLenght: dataSize,
+                position: offset,
+                bufferSize: bufferSize);
+
+            // Act
+            var readTest = new byte[bufferSize];
+            foreach (var _ in Enumerable.Range(0, 10))
+            {
+                await readStream.ReadAsync(readTest, 0, readTest.Length);
+                // seek back to exact start of current buffer
+                readStream.Position = offset;
+            }
+
+            // Assert
+            Assert.AreEqual(1, downloadMock.Invocations.Count);
```

Comment:
It might be good to include .Verify()'s somewhere, in addition to changing the mocking behavior to strict.